0016-primitive-options-type.md

```diff
@@ -0,0 +1,160 @@
+
+1. Use Python dataclass as the input/output type for primitive options. Dataclass allows auto-complete, and users can easily discover what fields are available.
+
+2. Make the `options` property callable, so a user can do `estimator.options.x`, to get/set the specific option, and `estimator.options()` to get all options. `options()` will return a copy of the options.
+
+2. `set_options` can continue to be used for bulk update.
+
+4. Providers that have more complex options, such as Qiskit Runtime, can use more advanced dataclass type such as the one in `pydantic`. `pydantic` dataclasses are very similar to Python dataclasses but allow easier validation. It would also return `True` when using `dataclasses.is_dataclass` to check its type.
+
+    While it is still up to each provider to implement the options, this allows a standardized `x.y` approach instead of separate `nnn_options` dictionaries. We can document the common grouping of options in base primitives, such as `transpilation` and `execution`, to encourage providers to use the same model.
+
+    Similarly, it is up to each provider on how they want to support undocumented/experimental options. Qiskit Runtime, for example, plans to support them under an `experimental` field that takes arbitrary kwargs.
+
+
+Backward compatibility depends on how we do migration for the primitive interface changes (see #51). If we decide to use a different version, then no backward compatibility is required. Otherwise, we'd need to use a different attribute, such as `settings`, and `options` would be deprecated and continue to return `qiskit.providers.Options`.
+
+## Detailed Design
+
+- `BasePrimitive` would accept either a dictionary or a dataclass for `options`.
+- The `options` property becomes callable, which would allow both auto-complete and getting a copy of all options.
+    - Alternatively, we can use `get_options` to return a copy of the options.
+- `set_options` can continue to be used for bulk update:
+
+```python
+from dataclasses import dataclass, replace
+
+@dataclass
+class BasePrimitiveOptions:
+
+    test_attr: int = 1
+
+    # Some recommended fields:
+    # transpilation: TranspilationOptions  # For transpiler options
+    # execution: ExecutionOptions  # For run options
+    # simualtor: SimulatorOptions  # For simulator options
+
+    def __call__(self):
+        return replace(self)
+
+
+class BasePrimitive:
+    """Primitive abstract base class."""
+
+    _options_class = BasePrimitiveOptions
+
+    def __init__(self, options: dict | BasePrimitiveOptions | None = None):
+        if isinstance(options, dict):
+          self._options = self._options_class(**options)
+        elif options is None:
+          self._options = self._options_class()
+        else:
+          self._options = options
+
+    @property
+    def options(self) -> BasePrimitiveOptions:
+        return self._options
+
+    def set_options(self, **fields):
+        self._options = replace(self._options, **fields)
```

Comment:
It seems silly to have a setter for the contents of a mutable object. I'm very in favour of just having the attribute exposed.

Reply:
So something like `estimator.options.update()` ? Makes sense to me.